go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -2158,3 +2166,106 @@ func generateIPFragmentReassemblyFlow(ofPortPhys string) []string {
 
 	return flows
 }
+
+// deleteStaleMasqueradeResources removes stale Linux resources when config.Gateway.V4MasqueradeSubnet
+// or config.Gateway.V6MasqueradeSubnet gets changed at day 2.
+func deleteStaleMasqueradeResources(bridgeName string) error {
+	var staleMasqueradeIPs config.MasqueradeIPsConfig
+	if config.Gateway.V4MasqueradeSubnet != config.Gateway.V4StaleMasqueradeSubnet {
+		if err := config.AllocateV4MasqueradeIPs(iputils.ParseIP(config.Gateway.V4StaleMasqueradeSubnet).ToIP(), &staleMasqueradeIPs); err != nil {
+			return fmt.Errorf("unable to determine stale V4MasqueradeIPs: %s", err)
+		}
+	}
+	if config.Gateway.V6MasqueradeSubnet != config.Gateway.V6StaleMasqueradeSubnet {
+		if err := config.AllocateV6MasqueradeIPs(iputils.ParseIP(config.Gateway.V6StaleMasqueradeSubnet).ToIP(), &staleMasqueradeIPs); err != nil {
+			return fmt.Errorf("unable to determine stale V6MasqueradeIPs: %s", err)
+		}
+	}
+	link, err := util.LinkSetUp(bridgeName)
+	if err != nil {
+		return fmt.Errorf("unable to get link for %s, error: %v", bridgeName, err)
+	}
+
+	if staleMasqueradeIPs.V4HostMasqueradeIP != nil || staleMasqueradeIPs.V6HostMasqueradeIP != nil {
+		if err = deleteMasqueradeResources(link, &staleMasqueradeIPs); err != nil {
+			return fmt.Errorf("unable to delete masquerade resources: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// deleteMasqueradeResources removes following Linux resources given a config.MasqueradeIPsConfig
+// struct and netlink.Link:
+// - neighbour object for IPv4 and IPv6 OVNMasqueradeIP and DummyNextHopMasqueradeIP.
+// - masquerade route added by addMasqueradeRoute function while starting up the gateway.
+// - iptables rules created for masquerade subnet based on ipForwarding and Gateway mode.
+// - stale HostMasqueradeIP address from gateway bridge
+func deleteMasqueradeResources(link netlink.Link, staleMasqueradeIPs *config.MasqueradeIPsConfig) error {
+	var subnets []*net.IPNet
+	var neighborIPs []net.IP
+	if config.IPv4Mode && staleMasqueradeIPs.V4HostMasqueradeIP != nil {
+		// Delete any stale masquerade IP from external bridge.
+		hostMasqIPNet, err := util.LinkAddrGetIPNet(link, staleMasqueradeIPs.V4HostMasqueradeIP)
+		if err != nil {
+			return fmt.Errorf("unable to get IPNet from link %s: %v", link, err)
+		}
+		if hostMasqIPNet != nil {
+			if err := util.LinkAddrDel(link, hostMasqIPNet); err != nil {
+				return fmt.Errorf("failed to remove masquerade IP from bridge %s: %w", link, err)
+			}
+		} else {
+			// This is to make sure that we don't delete anything if stale HostMasqueradeIP does not
```

Comment:
Instead of doing an early exit on errors/missing entries can we go through all cleanup steps and return one aggregated error? This would ensure that we would at least try to cleanup everything.

Reply:
How does returning nil ensures we are not removing user configured resources? This function assumes we own all of the resources that are being removed.

Reply:
I think I should get rid of this else condition. another downfall of this is that IPv6 masquerade resource cleanup may get hampered due to this return when IPv4 masquerade link address does not exist. 

So, I am going to aggregate all errors instead of returning immediately and going to log a warning message as Tim suggested.

Reply:
on a second thought it seems it would be better to have two different functions to delete v4 masquerade stale resources and v6 masquerade stale resources. I think its better to not delete some resources instead of deleting resources which user might have configured.
deleteV4MasqueradeResources & deleteV6MasqueradeResources: 
- If masquerade IP does not exist on link then dont remove other linux resources and just return nil error.
- Log warning message to notify user that some linux resources from stale masquerade subnet might still be there.
- Use aggregate type error so that the function does not return immediately upon a single failure while deleting resource.

Does this sound an acceptable option?

Reply:
masquerade IP does not exist on the *mp0* link right? I think this is preferable to me.

I'm a bit worried about this case:
1. A user is already using OVNK before this patch, and they use 169.254.169.x as a secondary IP address on their br-ex(breth0) interface already, so they were overriding masquerade subnet with something else.
2. The user upgrades to this new version, and we wipe his config due to the default value in v4StaleMasquerade.

I think by checking to see is the ip first on mp0, that gives us a signal that ok this was configured as a masquerade subnet before by ovnk, and gives us some validation that it should be ok for us to remove things related to that subnet. I think if it isn't there we cannot be sure that we should remove config and should log a warning. Its ok to do it independently for v4 and v6. I'm not sure I like the idea of propagating an error, unless it is just logged.